# Retail Sales Performance Analysis

## Business Objective

This project analyses retail sales performance to identify key revenue contributors, high-value customers, and strong-performing markets.

The analysis aims to provide data-driven insights that can support product, customer, and marketing decisions.

## Tools

- Python
- Pandas
- Jupyter Notebook

## Dataset Overview

The dataset contains three related tables:

- **Customers** — customer information and city
- **Orders** — order transactions and quantities
- **Products** — product information and prices

The tables are combined using CustomerID and ProductID to create an analysis-ready dataset.

In [10]:
import pandas as pd

customers = pd.DataFrame({
    'CustomerID': [1, 2, 3, 4],
    'CustomerName': ['Alice', 'Bob', 'Charlie', 'David'],
    'City': ['London', 'Manchester', 'Leeds', 'Bristol']
})

products = pd.DataFrame({
    'ProductID': [101, 102, 103],
    'ProductName': ['Laptop', 'Mouse', 'Keyboard'],
    'Price': [900, 25, 60]
})

orders = pd.DataFrame({
    'OrderID': [1, 2, 3, 4],
    'CustomerID': [1, 1, 2, 3],
    'ProductID': [101, 102, 103, 102],
    'Quantity': [1, 2, 1, 3]
})

## Data Quality Checks
Before analysing the data, I checked the dataset structure, missing values, duplicate records, and basic data validity.

In [11]:
# Check dataset structure
print('Customers:', customers.shape)
print('Orders:', orders.shape)
print('Products:', products.shape)

# Check missing values
print(customers.isna().sum())
print(orders.isna().sum())
print(products.isna().sum())

# Check duplicates
print('Duplicate customers:', customers.duplicated().sum())
print('Duplicate orders:', orders.duplicated().sum())
print('Duplicate products:', products.duplicated().sum())

Customers: (4, 3)
Orders: (4, 4)
Products: (3, 3)
CustomerID      0
CustomerName    0
City            0
dtype: int64
OrderID       0
CustomerID    0
ProductID     0
Quantity      0
dtype: int64
ProductID      0
ProductName    0
Price          0
dtype: int64
Duplicate customers: 0
Duplicate orders: 0
Duplicate products: 0


## Data Preparation

The Orders table was merged with Customers and Products to combine transaction, customer, and product information.

Sales was calculated as:

**Sales = Quantity × Price**

In [12]:
df = orders.merge(customers, on = 'CustomerID', how = 'left')
df = df.merge(products, on = 'ProductID', how = 'left')

#create sale column
df['Sales'] = df['Quantity'] * df['Price']
df

,OrderID,CustomerID,ProductID,Quantity,CustomerName,City,ProductName,Price,Sales
0,1,1,101,1,Alice,London,Laptop,900,900
1,2,1,102,2,Alice,London,Mouse,25,50
2,3,2,103,1,Bob,Manchester,Keyboard,60,60
3,4,3,102,3,Charlie,Leeds,Mouse,25,75


## Product Analysis
This analysis identifies which products contribute most to sales and which products have the highest unit volume.

In [13]:
product_analysis = (
    df.groupby('ProductName')
      .agg(
          TotalQuantity = ('Quantity', 'sum'),
          TotalSales = ('Sales', 'sum')
      )
      .reset_index()
)

product_analysis.sort_values(
    'TotalSales',
    ascending = False
)

,ProductName,TotalQuantity,TotalSales
1,Laptop,1,900
2,Mouse,5,125
0,Keyboard,1,60


### Finding
Laptop generated the highest revenue at £900, representing approximately 83% of total revenue in this dataset.

## Customer Analysis
Customer-level analysis was performed to identify high-value customers based on total spending, order volume, and quantity purchased.

In [14]:
customer_analysis = (
    df.groupby('CustomerName')
    .agg(
        TotalOrders = ('OrderID', 'count'),
        TotalQuantity = ('Quantity', 'sum'),
        TotalSales = ('Sales', 'sum')
    )
    .reset_index()
)

customer_analysis.sort_values(
    'TotalSales',
    ascending = False
)

,CustomerName,TotalOrders,TotalQuantity,TotalSales
0,Alice,2,3,950
2,Charlie,1,3,75
1,Bob,1,1,60


### Finding
Alice was the highest-value customer by total spending, generating £950 and contributing approximately 88% of total revenue.

## City Analysis
City-level analysis was used to identify markets with the highest sales revenue and sales volume.

In [15]:
city_analysis = (
    df.groupby('City')
    .agg(
        TotalQuantity = ('Quantity', 'sum'),
        TotalSales = ('Sales', 'sum')
    )
    .reset_index()
)

city_analysis.sort_values(
    'TotalSales',
    ascending = False
)

,City,TotalQuantity,TotalSales
1,London,3,950
0,Leeds,3,75
2,Manchester,1,60


### Finding
London generated the highest sales revenue at £950, making it the strongest-performing market in the current dataset.

## Customer Revenue Contribution
Revenue share was calculated to understand how total sales are distributed across customers.

In [16]:
customer_analysis['RevenueShare'] = (
    customer_analysis['TotalSales'] / customer_analysis['TotalSales'].sum() * 100
)

customer_analysis.sort_values(
    'RevenueShare',
    ascending = False
)

,CustomerName,TotalOrders,TotalQuantity,TotalSales,RevenueShare
0,Alice,2,3,950,87.557604
2,Charlie,1,3,75,6.912442
1,Bob,1,1,60,5.529954


### Finding
Alice contributed approximately 88% of total revenue in the current dataset, indicating a high concentration of revenue among a single customer.

## Key Insights
1. **Product:** Laptop generated £900 in revenue, representing approximately 83% of total revenue and making it the main revenue contributor in this dataset.
2. **Customer:** Alice was the highest-value customer by total spending, generating £950 and contributing approximately 88% of total revenue.
3. **Market:** London generated £950 in sales revenue, making it the strongest-performing market in the current dataset.

## Recommendations
1. Consider maintaining strong product availability for laptops and exploring targeted marketing opportunities for this product category.
2. Consider introducing a loyalty programme for high-value customers based on defined spending criteria, with benefits such as early access to products or special offers.
3. Consider further investigating London as a target market and testing targeted local marketing campaigns before making larger investment decisions.

## Conclusion
The analysis identified the main revenue contributors across products, customers, and markets. Laptop, Alice, and London were the strongest performers in their respective areas.
However, the dataset is small and does not contain historical trends, profitability, marketing costs, or broader market information. Further analysis would therefore be required before making larger strategic decisions.